### PR TITLE
fix: template with no output should create an empty output dir

### PR DESCRIPTION
**Pull request by drevell**
The former behavior was that the output directory would not be created
in the edge case where the template produced no output files. This was
problematic for two reasons:

 - It broke `golden-test record`; it expected to find an output directory
 - Empty output is semantically valid. Suppose there's a template that
   has a bunch of optional steps that are all skipped.

This requires changes to template rendering, golden-test recording, and golden-test verification.

I also decided to use the `.abc` directory name for both manifests and golden-test recording metadata. It's simple and consistent to have just one metadata directory name that we use everywhere.

### CODE REVIEW

File: templates/commands/goldentest/test_funcs.go [hunk 1]
```diff
@@ -156,10 +156,6 @@ func renderTestCases(ctx context.Context, testCases []*TestCase, location string
 func renderTestCase(ctx context.Context, templateDir, outputDir string, tc *TestCase) error {
 	testDir := filepath.Join(outputDir, goldenTestDir, tc.TestName, testDataDir)
 
-	if err := os.RemoveAll(testDir); err != nil {
```
Comment (capri-xiyue) [inline templates/commands/goldentest/test_funcs.go:159]: Is this the wrong logic that we want to remove?

File: templates/common/paths.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+package common
+
+import (
+	"path/filepath"
+	"strings"
+)
+
+// ABCInternalDir is the name for internal directories that have things like
+// manifests and .gitkeep files. These may be destination directories or
+// golden-test data/ directories.
```
Comment (capri-xiyue) [inline templates/common/paths.go:24]: nits is this a typo for `golden-test data/`

File: templates/commands/goldentest/test_funcs.go (hunk 1 above)
Comment (drevell) [reply templates/commands/goldentest/test_funcs.go:159]: This was supposed to be deleted in a previous PR but seems to have been lost in a merge conflict, so I'm deleting it again

File: templates/common/paths.go (hunk 2 above)
Comment (drevell) [reply templates/common/paths.go:24]: I think it's good as written

File: templates/commands/goldentest/record.go [hunk 3]
```diff
@@ -112,6 +112,16 @@ func (c *RecordCommand) Run(ctx context.Context, args []string) error {
...
 		}
 		merr = errors.Join(merr, common.CopyRecursive(ctx, nil, params))
+
+		abcInternal := filepath.Join(testDir, common.ABCInternalDir)
+		if err := os.MkdirAll(abcInternal, common.OwnerRWXPerms); err != nil {
+			return fmt.Errorf("failed to create dir %q: %w", abcInternal, err)
+		}
+		// git won't commit an empty directory, so add a placeholder file.
+		gitKeep := filepath.Join(abcInternal, ".gitkeep")
+		if err := os.WriteFile(gitKeep, []byte{}, common.OwnerRWPerms); err != nil {
```
Comment (sethvargo) [inline templates/commands/goldentest/record.go:122]: I believe you can pass `nil` in for the content.

**Review by sethvargo — APPROVED** [review]

File: templates/commands/goldentest/record.go (hunk 3 above)
Comment (drevell) [reply templates/commands/goldentest/record.go:122]: Merging now to avoid the velocity hit for another round of review, will fix in a separate PR.
